src/daemon/thread_utils.c

```diff
@@ -0,0 +1,120 @@
+#include <daemon/thread_utils.h>
+#include <errno.h>
+#include <fts.h>
+#include <libgen.h>
+#include <sched.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+struct Job *find_job_by_id(int64_t id) {
+    if (id >= 0 && id < job_count)
+        return jobs[id];
+    return NULL;
+}
+
+void check_suspend(struct Job *job_to_check) {
+    pthread_mutex_t *status_mutex = &job_to_check->status_mutex;
+    pthread_cond_t *resume_cond = &job_to_check->mutex_resume_cond;
+
+    pthread_mutex_lock(status_mutex);
+    while (job_to_check->status == JOB_STATUS_PAUSED)
+        pthread_cond_wait(resume_cond, status_mutex);
+    pthread_mutex_unlock(status_mutex);
+}
+
+/**
+ * Creates a new directory
+ * path - the path of the new directory
+ * parent - the parent of the new directory, 
+ * should be NULL if the new directory is root
+ */
+struct Directory *create_directory(char *path, struct Directory *parent) {
+    struct Directory *new_directory = malloc(sizeof(struct Directory));
+    new_directory->path = malloc((strlen(path) + 1) * sizeof(char));
+    strcpy(new_directory->path, path);
+    new_directory->parent = parent;
+    new_directory->bytes = new_directory->number_files = 0;
+    new_directory->children_capacity = new_directory->number_subdir = 0;
+    new_directory->children = NULL;
+
+    if (parent != NULL) {
+      // Add the new directory to the list of his parents children
+      if (parent->children_capacity == parent->number_subdir) {
+        if (parent->children_capacity == 0)
+            parent->children_capacity = 1;
+        else
+            parent->children_capacity *= 2;
+
+        parent->children = realloc(parent->children, parent->children_capacity);
```

Comment:
Should be `realloc(parent->children, parent->children_capacity * sizeof(struct Directory *));`